### PR TITLE
Stargate msg and query

### LINKED ISSUE #388: Add support for Stargate CosmosMsg/QueryRequest
In master branch (v0.14-dev) of CosmWasm, we add a type for dispatching a protobuf-encoded "stargate message", or making a protobuf-encoded query to allow quick use of any mdoule that doesn't have a custom handler.

This is exposed in `wasmvm` go types and the most recent `reflect` contract, but we need to add dispatch support to `wasmd` and test it.

### PR DESCRIPTION
**Pull request by ethanfrey**
Closes #388 

- [x] Implement StargateMsgEncoder
- [x] Implement StargateQueryHandler
- [x] Test message encoder
- [x] Test query handler

### CODE REVIEW

File: x/wasm/internal/keeper/handler_plugin.go [hunk 1]
```diff
@@ -20,8 +22,8 @@ type DefaultMessageHandler struct {
 	encoders MessageEncoders
 }
 
-func NewDefaultMessageHandler(router sdk.Router, channelKeeper types.ChannelKeeper, capabilityKeeper types.CapabilityKeeper, customEncoders *MessageEncoders) DefaultMessageHandler {
-	encoders := DefaultEncoders(channelKeeper, capabilityKeeper).Merge(customEncoders)
+func NewDefaultMessageHandler(router sdk.Router, channelKeeper types.ChannelKeeper, capabilityKeeper types.CapabilityKeeper, unpacker codectypes.AnyUnpacker, customEncoders *MessageEncoders) DefaultMessageHandler {
```
Comment (alpe) [inline x/wasm/internal/keeper/handler_plugin.go:25]: good use of the minimal interface

File: x/wasm/internal/keeper/handler_plugin.go [hunk 2]
```diff
@@ -170,6 +180,20 @@ func EncodeStakingMsg(sender sdk.AccAddress, msg *wasmvmtypes.StakingMsg) ([]sdk
...
+	return func(sender sdk.AccAddress, msg *wasmvmtypes.StargateMsg) ([]sdk.Msg, error) {
+		any := codectypes.Any{
+			TypeUrl: msg.TypeURL,
+			Value:   msg.Value,
+		}
+		var sdkMsg sdk.Msg
+		if err := unpacker.UnpackAny(&any, &sdkMsg); err != nil {
+			return nil, sdkerrors.Wrap(types.ErrInvalidMsg, fmt.Sprintf("Cannot unpack proto message with type URL: %s", msg.TypeURL))
+		}
+		return []sdk.Msg{sdkMsg}, nil
```
Comment (alpe) [inline x/wasm/internal/keeper/handler_plugin.go:193]: The `sdkMsg` can contain `Any` types as well that need to be unpacked. This is not done recursively (https://github.com/cosmos/cosmos-sdk/blob/v0.41.4/codec/types/compat.go#L197).
You need another call to `codectypes.UnpackInterfaces(sdkMsg, unpacker)`

Comment (ethanfrey) [reply x/wasm/internal/keeper/handler_plugin.go:196]: I did this here: https://github.com/CosmWasm/wasmd/pull/435/commits/c33b3694acef890e5754e2e0e75ac0bf9a701a14

Can you check that was correct?

Comment (alpe) [reply x/wasm/internal/keeper/handler_plugin.go:196]: ja, looks good!
btw. you can use `sdkerrors.Wrapf` to save you some `fmt.Sprintf`

File: x/wasm/internal/keeper/handler_plugin_test.go [hunk 3]
```diff
@@ -276,6 +288,27 @@ func TestEncoding(t *testing.T) {
 				},
 			},
 		},
+		// TODO: alpe? can you add an example with sub-interfaces (where the UnpackInterfaces call would be needed)
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/handler_plugin_test.go:291]: I am not sure how to create such a case. (I also think we only have bank, staking, and distribution registered here in the tests).

If you could add such a test to trigger this that would be great. Or if it is a pain, please open up an issue on it and explain more or less what is needed. I can put that down lower on the priority list

**Review by ethanfrey — COMMENTED** [review]
@alpe this should be ready now. I'd like a review, especially on the proto part and how I wire it with the app

File: x/wasm/internal/keeper/handler_plugin_test.go (hunk 3 above)
Comment (alpe) [reply x/wasm/internal/keeper/handler_plugin_test.go:291]: I found `Any` types in sdk (master). 
* MsgGrantAuthorizationRequest
* MsgGrantFeeAllowance
* MsgSubmitProposal
* MsgCreateValidator
I guess `MsgSubmitProposal` would be a good candidate. I can add a test.

File: x/wasm/internal/keeper/reflect_test.go [hunk 4]
```diff
@@ -296,6 +297,79 @@ func TestMaskReflectCustomQuery(t *testing.T) {
...
+	mustParse(t, simpleChain.Data, &simpleBalance)
+	require.Equal(t, len(expectedBalance), len(simpleBalance.Amount))
+	assert.Equal(t, simpleBalance.Amount[0].Amount, expectedBalance[0].Amount.String())
+	assert.Equal(t, simpleBalance.Amount[0].Denom, expectedBalance[0].Denom)
+
+	// now, try to build a protobuf query
+	protoQuery := banktypes.QueryAllBalancesRequest{
+		Address: creator.String(),
+	}
+	protoQueryBin, err := proto.Marshal(&protoQuery)
```
Comment (alpe) [inline x/wasm/internal/keeper/reflect_test.go:348]: Clever to test with a different golang protobuf marshaller.  🌻

File: x/wasm/internal/keeper/reflect_test.go [hunk 5]
```diff
@@ -296,6 +297,79 @@ func TestMaskReflectCustomQuery(t *testing.T) {
...
+	})
+	require.NoError(t, err)
+	simpleRes, err := keeper.QuerySmart(ctx, contractAddr, simpleQueryBz)
+	require.NoError(t, err)
+	var simpleChain ChainResponse
+	mustParse(t, simpleRes, &simpleChain)
+	var simpleBalance wasmvmtypes.AllBalancesResponse
+	mustParse(t, simpleChain.Data, &simpleBalance)
+	require.Equal(t, len(expectedBalance), len(simpleBalance.Amount))
+	assert.Equal(t, simpleBalance.Amount[0].Amount, expectedBalance[0].Amount.String())
```
Comment (alpe) [inline x/wasm/internal/keeper/reflect_test.go:341]: nit: no need to change: wrong order of arguments. Here and others (t, expect, got)  (to be fair it is not consistent within `testify` like assert.Len(t, got, exp)` )

**Review by alpe — APPROVED** [review]
Great work. This will add a very valuable feature to the system. 🚀

**Comment by ethanfrey** [comment]
Thanks. If you could add a follow up pr sometime with submit proposal, that would be great 

Now contracts can vote

**Pull request merged**